### ISSUE TITLE
Sharing via Email + password protected link --> File not found

### ISSUE THREAD

**Issue by Perry2493** [nextcloud/server#13885]
Hi Guys,

if have a problem when shareing files and/or directories via email.
After upgrading from NC12 to NC13 and then to NC14 i can't share anything via email anymore.
The recipient will get both emails (link to file/directory + password) but when opening that link he gets an error "File not Found".
I am running the App FileAccessControl to permit acces only from internal IP's or on files which are tagged for external shareing, so i thought that could be the issue - but even after disabling that App I still have that problem.

It is working, when i share a file with another User of my NextCloud (in that case even the link in Email works)- my problem only occurs when sharing via email to someone that has no user in my NC.

### Steps to reproduce
1.   Share File or Directory via Email
2.   Click Link in Email

### Expected behaviour
NC should Ask for Password for that Share/File

### Actual behaviour
Error "File not Found"

### Server configuration

**Operating system**: Ubuntu 16.04.5 LTS

**Web server:** Apache 2.4.18

**Database:** MariaDB 10.0.36

**PHP version:** 7.0

**Nextcloud version:** 14.0.6

**Updated from an older Nextcloud/ownCloud or fresh install:** Updated from 12.0.x to 13.0.x and then to 14.0.6

**Where did you install Nextcloud from:** website

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.0.1
  - activity: 2.7.0
  - admin_audit: 1.4.0
  - bruteforcesettings: 1.3.0
  - cloud_federation_api: 0.0.1
  - comments: 1.4.0
  - dav: 1.6.1
  - encryption: 2.2.0
  - federatedfilesharing: 1.4.0
  - federation: 1.4.0
  - files: 1.9.0
  - files_accesscontrol: 1.4.0
  - files_automatedtagging: 1.4.0
  - files_pdfviewer: 1.3.2
  - files_sharing: 1.6.2
  - files_texteditor: 2.6.0
  - files_trashbin: 1.4.1
  - files_versions: 1.7.1
  - files_videoplayer: 1.3.0
  - firstrunwizard: 2.3.0
  - gallery: 18.1.0
  - groupfolders: 2.0.2
  - logreader: 2.0.0
  - lookup_server_connector: 1.2.0
  - nextcloud_announcements: 1.3.0
  - notifications: 2.2.1
  - oauth2: 1.2.1
  - onlyoffice: 2.1.2
  - password_policy: 1.4.0
  - provisioning_api: 1.4.0
  - serverinfo: 1.4.0
  - sharebymail: 1.4.0
  - support: 1.0.0
  - survey_client: 1.2.0
  - systemtags: 1.4.0
  - theming: 1.5.0
  - twofactor_backupcodes: 1.3.1
  - updatenotification: 1.4.1
  - user_ldap: 1.4.0
  - workflowengine: 1.4.0
Disabled:
  - announcementcenter
  - files_external
  - spreedme
  - user_external
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "my.internal.domain.blah",
            "my.domain.blah"
        ],
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "forwarded_for_headers": [
            "HTTP_X_FORWARDED_FOR",
            "HTTP_FORWARDED_FOR"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "https:\/\/my.domain.at",
        "enable_previews": false,
        "dbtype": "mysql",
        "version": "14.0.6.0",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "activity_expire_days": 180,
        "preview_max_x": "960",
        "preview_max_y": "720",
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "\\OCA\\User_LDAP\\LDAPProviderFactory",
        "loglevel": 2,
        "logfile": "\/var\/log\/nextcloud.log",
        "maintenance": false,
        "singleuser": true,
        "updater.release.channel": "production",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "PLAIN",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "theme": "",
        "skeletondirectory": "",
        "filesystem_check_changes": 0,
        "updater.secret": "***REMOVED SENSITIVE VALUE***"
    }
}
```
</details>

**Are you using external storage, if yes which one:** nfs-mount from an netapp

**Are you using encryption:** yes

**Are you using an external user-backend, if yes which one:** ActiveDirectory

### Client configuration
**Browser:** Firefox/Chrome/Edge

**Operating system:** Win10

#### Web server error log
Actually no new entry in webserver error log when trying to access shared file/directory

#### Nextcloud log (data/nextcloud.log)
Actually no new entry in Nextcloud log when trying to access shared file/directory

Please let me know if you need some more informations from me.
Thanks for your help and best regards,
Perry

**Comment by Perry2493** [comment]
Hi.
i sadly still have no clue what could cause such behavior.

Every link works - even if i share a file to another user of my NC-instance, the link in his notificaton-email works.
Only if i E-Mail-Share something to someone who has no User in my NC the link in his email is broken and pointing to nothing...

How could that be possible? What am i missing?
Sorry for reposting, but this feature is highly demanded im my company :-/

Thanks for any help and best regards, 
Perry

**Comment by Perry2493** [comment]
Still have that Problem - still don't know what to do.
Guess it's no know issue in NC 14.0.x ?

Should i just upgrade to Version 15 and pray it'll fix that problem?

BR, Perry

**Comment by kungknut** [comment]
Unfortunately, I will have to second on this.

We are running NC15.0.5 and having the same problem.

**Comment by Perry2493** [comment]
Hi, 
i just upgraded to NC15.0.6 - problem still persists.
Any help is apreciated - thank you guys..

BR, Perry

**Comment by FlyinFistOfJuda** [comment]
If the problem still persists: It's based on the "Group Folders" App inf Apps/Files. Deactivate it (if possible) and your links shared by mail will work again. Tested on 15.0.7 & 16.0.1.

**Comment by Perry2493** [comment]
OK, i guess i found a solution - or at least a reason...
As long as "Allow users to share via link" is enabled, "Share by mail" is working without Problems.
As soon as "Allow users to share via link" is disabled, "Share by mail" isn't working anymore.

I mean yes, basically it is just a sharing-link in an email - but in older Versions of NC (at least at NC12) it was possible to disable link-sharing while having E-Mail-Sharing enabled. 
So i was able to force users to create a unique link for every share - which makes it easier to have a overview.

Just tested it on an NC16-Instance from one of my customers and on my own NC15-Instance.
Same behaviour on both Instances.

So i'm not sure - is it a bug or is it working as intended?
Is "Share by mail" seen as "Sharing via link"?

Thanks and BR,
Perry

**Comment by DavidDzl** [comment]
I can confirm the issue still exists in 16.0.3.

and @Perry2493 's suggestion works.

**Comment by skjnldsv** [comment]
Fixed with #15719 and the various recent fixes to the sharing api

Sharing via mail is the same as sharing via link, we now make sure to aply the same config and restrictions to it
